core/src/main/java/org/apache/calcite/rel/rel2sql/CTERelToSqlUtil.java

```diff
@@ -137,34 +175,137 @@ public static void addSqlWithItemNode(SqlWithItem sqlWithItem, List<SqlNode> sql
    */
   public static void updateSqlNode(SqlNode sqlNode) {
     if (sqlNode != null) {
-      SqlSelect sqlSelect = (SqlSelect) sqlNode;
-      if (sqlSelect.getFrom() instanceof SqlJoin) {
-        SqlNode leftNode = ((SqlJoin) sqlSelect.getFrom()).getLeft();
-        SqlNode rightNode = ((SqlJoin) sqlSelect.getFrom()).getRight();
-
-        //Update Left Node
-        updateNode(leftNode);
-        // Update Right Node
-        updateNode(rightNode);
-
-        //if left and right is join again need recursive call
-      } else if (sqlSelect.getFrom() instanceof SqlBasicCall) {
-        updateNode(sqlSelect.getFrom());
-      } else if (sqlSelect.getFrom() instanceof SqlWithItem) {
-        sqlSelect.setFrom(((SqlWithItem) sqlSelect.getFrom()).name);
+      if (sqlNode instanceof SqlSelect) {
+        SqlSelect sqlSelect = (SqlSelect) sqlNode;
+        // Handle FROM clause
+        SqlNode fromNode = sqlSelect.getFrom();
+        processFromNode(sqlSelect, fromNode);
+        if (isNestedCte(fromNode)
+            &&
+            fromNode instanceof SqlBasicCall
+            &&
+            ((SqlBasicCall) fromNode).getOperator() instanceof SqlAsOperator) {
+          updateNode(fromNode);
+        }
+        // Handle WHERE clause
+        SqlNode whereNode = sqlSelect.getWhere();
+        if (whereNode != null && whereNode instanceof SqlBasicCall) {
```

Comment:
Will this check `whereNode instanceof SqlBasicCall` not be sufficient?? Do we need null check